### PR TITLE
Part3 - Post handling

### PR DESCRIPTION
**Pull request by tidyjava**

### CODE REVIEW

File: src/test/groovy/com/tidyjava/bp/PostControllerSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    def "missing post"(n) {
+        expect:
+        mockMvc.perform(get("/surely-not-existent"))
+                .andExpect(status().isNotFound())
+                .andExpect(view().name("not-found"))
+
+        where:
+        n << [1, 2]
```
Comment (tidyjava) [inline src/test/groovy/com/tidyjava/bp/PostControllerSpec.groovy]: copy-paste leftover

File: src/test/resources/application.yml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+blog:
```
Comment (tidyjava) [inline src/test/resources/application.yml]: is this file needed at all?

File: src/test/groovy/com/tidyjava/bp/PostControllerSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    MockMvc mockMvc
+
+    def setup() {
+        mockMvc = MockMvcBuilders.webAppContextSetup(this.wac).build()
+    }
+
+    def "home"() {
+        expect:
+        def result = mockMvc.perform(get("/"))
+                .andExpect(status().isOk())
```
Comment (tidyjava) [inline src/test/groovy/com/tidyjava/bp/PostControllerSpec.groovy]: expect a view

File: src/test/groovy/com/tidyjava/bp/PostControllerSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                .andReturn()
+        def posts = result.modelAndView.model.posts
+        assertTestPost(posts[0], 2)
+        assertTestPost(posts[1], 1)
+    }
+
+    def "post"(n) {
+        expect:
+        def result = mockMvc.perform(get("/post$n"))
+                .andExpect(status().isOk())
```
Comment (tidyjava) [inline src/test/groovy/com/tidyjava/bp/PostControllerSpec.groovy]: expect a view

File: src/test/groovy/com/tidyjava/bp/PostReaderSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+import spock.lang.Specification
+
+class PostReaderSpec extends Specification {
+
+    def postReader = new PostReader()
+
+    def "no posts"() {
+        given:
+        postReader.postsLocation = "non-existent-or-empty"
```
Comment (tidyjava) [inline src/test/groovy/com/tidyjava/bp/PostReaderSpec.groovy]: this is non-existent. add empty

File: src/test/groovy/com/tidyjava/bp/PostReaderSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package com.tidyjava.bp
```
Comment (darzieba) [inline src/test/groovy/com/tidyjava/bp/PostReaderSpec.groovy:1]: No tests for happy path?

File: src/test/groovy/com/tidyjava/bp/PostControllerSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import spock.lang.Specification
+
+import static java.time.format.DateTimeFormatter.ISO_LOCAL_DATE
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.view
+
+@SpringApplicationConfiguration(BloggingPlatform.class)
+@WebAppConfiguration
+class PostControllerSpec extends Specification {
```
Comment (darzieba) [inline src/test/groovy/com/tidyjava/bp/PostControllerSpec.groovy]: If this is integration test, maybe it will be good to name class like PostControllerIntegrationSpec?

File: src/main/java/com/tidyjava/bp/PostReader.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+import static java.util.stream.Collectors.toList;
+
+@Service
+public class PostReader {
+
+    @Value("${posts.location}")
+    private String postsLocation;
+
+    public List<MarkdownPost> readAll() {
+        URL postsRes = getResource(postsLocation);
```
Comment (darzieba) [inline src/main/java/com/tidyjava/bp/PostReader.java:24]: I don't like those shortened names everywhere, makes code less readable

File: src/main/java/com/tidyjava/bp/PostReader.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        if (notExists(postsRes)) {
+            return emptyList();
+        }
+        File postsDir = toFile(postsRes);
```
Comment (darzieba) [inline src/main/java/com/tidyjava/bp/PostReader.java:28]: The logic from 25-28 doesn't need to be executed with every request, it won't change during service lifecycle

File: src/main/java/com/tidyjava/bp/MarkdownPost.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+import static java.util.Collections.singletonList;
+
+public class MarkdownPost {
+    public static final String EXTENSION = ".md";
+
+    private Node parsedFile;
+    private Map<String, List<String>> metadata;
+    private String url;
+
+    public MarkdownPost(File file) {
```
Comment (darzieba) [inline src/main/java/com/tidyjava/bp/MarkdownPost.java:26]: a lot of methods/constructors in this package can be made package local

Comment (darzieba) [inline src/main/java/com/tidyjava/bp/MarkdownPost.java:26]: is constructor really a good place for it? not intuitive for sure...
What about MarkdownPost.from(file)?
